### PR TITLE
fix: bug that resulted in some deaths getting counted twice

### PR DESCRIPTION
**Pull request by RobertB4**
## Problem
The `PLAYER_DEAD` event triggers when:
- A player dies
- A player logs out after dying without releasing their spirit and logging back in

This leads to some deaths getting counted twice.

## Solution
In addition to updating deaths, I added an additional check using the `PLAYER_ALIVE` event. This event triggers when:
- A player releases their spirit (either by pressing the button, or automatically after logging in dead after they logged out without releasing their spirit).
- There are other cases where this event triggers, e.g. on every log in. Unfortunately there is no complete documentation on this 😕 

From my understanding and testing, it looks like every `PLAYER_DEAD` event is followed by exactly one `PLAYER_ALIVE` event.
- Either because the player pressed the release spirit button
- Or because the player logged out before releasing and logged back in

My solution is to "verify" a death using the `PLAYER_ALIVE` event, meaning when 2 consecutive  `PLAYER_DEAD` events are fired without having a `PLAYER_ALIVE` event in between, the second `PLAYER_DEAD` does not update deaths.

## Tests
The following cases work, meaning they only count as one death:
- Die, logout before releasing, log back in
- Die, release spirit
- Die, release spirit, reload UI while dead
- Die, release spirit, logout and back in

Not sure if there are any other potential edge cases we should test. Feedback appreciated!

### CODE REVIEW

File: Hardcore.lua [hunk 1]
```diff
@@ -250,7 +251,7 @@ function Hardcore:PLAYER_LOGIN()
 			time_played = 0,
 			accumulated_time_diff = 0,
 			tracked_played_percentage = 0,
-			deaths = 0,
+			deaths = {},
```
Comment (aaronma37) [inline Hardcore.lua:254]: I like this solution! I am concerned about future-proofing edits to Hardcore_Character types.  Does this cause an error down the line if we are trying to index based on player_alive_trigger etc..

Comment (RobertB4) [reply Hardcore.lua:254]: Not sure what you mean, do you have an example of what you think could go wrong?

Comment (aaronma37) [reply Hardcore.lua:254]: Yeah, its not really an issue at the moment or with this PR but if we have version releases where Hardcore_Character.var changes from an int to a dictionary, that var will remain as an int since its already initialized. Later code paths might try to use that var as a dict, Hardcore_Character.var.key.

Comment (RobertB4) [reply Hardcore.lua:254]: Ok I understand what you mean now!
If we want to change the data type of saved variables, we have to add a legacy case to handle the old data type.
For example in this case, we could check if Harcore_Character.deaths is of type number, and if it is we create a table with n records where n is the number saved in Hardcore_Character.deaths, and only run code that requires deaths to be a table after.

Might actually be worth adding it even for this, even though it is not an issue on release, but the streamers who have been testing the addon will get an error if they log onto characters that have deaths saved as a number.

Comment (zdeyn) [reply Hardcore.lua:254]: It's pretty straight-forward to put data-type cleaning as part of our overall library architecture - read cleanly as both old and new types, but always write new type

**Review by zdeyn — APPROVED** [review]
If this works as intended for now, we can solve it properly down the track with variable access abstraction in the utils/library

Box it and ship it

File: Hardcore.lua (hunk 1 above)
Comment (aaronma37) [reply Hardcore.lua:254]: Seems like potentially a lot of cases to cover. What if we used a more modular structure in the future so we don't have to add bridges for releases? I think the dict struct that rei is using is pretty future proof

Comment (zdeyn) [reply Hardcore.lua:254]: any time we change a typed variable, we'd have to use a method, e.g. HCDB:SetValue('whatever', somevalue)
any time we want to use it, we'd use something different, e.g. HCDB:GetValue('whatever', optional_default)

the problem solves itself over time

**Comment by RobertB4** [comment]
Added handling for legacy deaths.

**Review by aaronma37 — APPROVED** [review]
